jmdaemon/jmdaemon/onionmc.py

```diff
@@ -1084,10 +1111,27 @@ def process_control_message(self, peerid: str, msgtype: int,
                 return True
             try:
                 peerlist = msgval.split(",")
-                for peer in peerlist:
+                for peer_in_list in peerlist:
+                    # directories should send us peerstrings that include
+                    # nick;host:port;D where "D" indicates that the directory
+                    # is signalling this peer as having left. Otherwise, without
+                    # the third field, we treat it as a "join" event.
+                    try:
+                        nick, hostport, disconnect_code = peer_in_list.split(
+                            NICK_PEERLOCATOR_SEPARATOR)
+                        if not disconnect_code == "D":
+                            continue
+                        self.on_nick_leave_directory(nick, peer)
+                        continue
+                    except ValueError:
+                        # old code does not recognize this "D"; it will
+                        # swallow the message in `add_peer`, ignoring
+                        # the message as invalid because it has three fields
+                        # instead of two.
+                        pass

```

Comment:
I might be confused, but:

* Isn't this comment out of place? Why not put this comment in, e.g., `send_peers`? This `except` is already covered by the comment right before saying `Otherwise, without the third field, we treat it as a "join" event`, right?

* Can this `except` be risky/too loose? `ValueError` might be triggered in other ways than the one we expect.

Reply:
> This except is already covered by the comment right before saying Otherwise, without the third field, we treat it as a "join" event, right?

Yes i suppose the comment is vaguely misplaced, but it's kind of misplaced anywhere(?), because it's a comment about what happens in the previous version of the code, to this. So .. probably better put it inside the function `add_peer`, in the first section that parses the message as having two fields.

Re: just `continue`-ing instead of raising an error if not "D", I was doing that as a forwards compatibility thing, so that if we choose to add a different code there, old versions will just ignore the message.

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1249/commits/706bdb0a0379ed6ed841ee02c7fc77dff0c22b7b : moved this comment to a more logical case, the `Exception` catch in `add_peer`.

Re: `ValueError`, I don't think it can be too loose since reaching `pass` there just results in us reverting to the non-"D" parsing in `self.add_peer()`.